### DISCUSSION TITLE
question about uv__tcp_connect

### DISCUSSION THREAD

**Discussion by theanarkh** [libuv/libuv#3903] [Q&A] [answered]
Hi, why the `uv__tcp_connect` still register the `POLLOUT` event even if `connect` function return error. the related code is as follows.
```
  uv__req_init(handle->loop, req, UV_CONNECT);
  req->cb = cb;
  req->handle = (uv_stream_t*) handle;
  QUEUE_INIT(&req->queue);
  handle->connect_req = req;

  uv__io_start(handle->loop, &handle->io_watcher, POLLOUT);

  if (handle->delayed_error)
    uv__io_feed(handle->loop, &handle->io_watcher);
```
but the `uv_pipe_connect` only register the `POLLOUT` event when `err == 0`.
```
  if (err == 0)
    uv__io_start(handle->loop, &handle->io_watcher, POLLOUT);

out:
  handle->delayed_error = err;
  handle->connect_req = req;

  uv__req_init(handle->loop, req, UV_CONNECT);
  req->handle = (uv_stream_t*)handle;
  req->cb = cb;
  QUEUE_INIT(&req->queue);

  /* Force callback to run on next tick in case of error. */
  if (err)
    uv__io_feed(handle->loop, &handle->io_watcher);

```

**Comment by bnoordhuis** [comment, marked as answer]
I agree it's a little inconsistent but it should make no difference to the observable behavior of the program. The `uv__io_start()` call in `uv__tcp_connect()` is superfluous, it's the `uv__io_feed()` call that drives the program forward.

The superfluous call only happens on OpenBSD and Solaris so yeah, I'm not going to lose sleep over that. :-)

**Comment by theanarkh** [comment]
it seems
1. `uv__io_start` will insert the handle into `loop->watcher_queue`.
2. then `uv__io_poll` will register the event to `epoll`.
3. finally `uv__stream_connect` call `uv__io_stop` to remove the `POLLOUT` event. 

Does this lead to some unnecessary operations ?

**Reply by bnoordhuis** [reply]
It could, but it probably won't. Pending I/O watchers (`uv__io_feed()` makes an I/O watcher pending) run before and after `uv__io_poll()` so most of the time they've already been dealt with before they make it to `uv__io_poll()`.

**Reply by theanarkh** [reply]
Get it, thank you very much.
